File: kotlin/src/Day02.kt
import java.lang.IllegalArgumentException
import java.lang.Long.parseLong

data class PositionWithAim(val position: Long, val depth: Long, val aim: Long)
data class Position(val position: Long, val depth: Long)

sealed class Command(protected val magnitude: Long) {
    abstract fun move(position: Position): Position
    abstract fun moveAimed(positionWithAim: PositionWithAim): PositionWithAim
}

class Forward(magnitude: Long): Command(magnitude) {
    override fun move(position: Position): Position {
        return Position(position.position + magnitude, position.depth)
    }

    override fun moveAimed(positionWithAim: PositionWithAim): PositionWithAim {
        return PositionWithAim(
            positionWithAim.position + magnitude,
            positionWithAim.depth + (positionWithAim.aim * magnitude),
            positionWithAim.aim
        )
    }
}

class Down(magnitude: Long): Command(magnitude) {
    override fun move(position: Position): Position {
        return Position(position.position, position.depth + magnitude)
    }

    override fun moveAimed(positionWithAim: PositionWithAim): PositionWithAim {
        return PositionWithAim(positionWithAim.position, positionWithAim.depth, positionWithAim.aim + magnitude)
    }
}

class Up(magnitude: Long): Command(magnitude) {
    override fun move(position: Position): Position {
        return Position(position.position, position.depth - magnitude)
    }

    override fun moveAimed(positionWithAim: PositionWithAim): PositionWithAim {
        return PositionWithAim(positionWithAim.position, positionWithAim.depth, positionWithAim.aim - magnitude)
    }
}

fun main() {
    fun part1(input: List<String>): Long {
        val finalPosition = input.map { line ->
            val command = line.split(" ")
            when (command.first()) {
                "forward" -> Forward(parseLong(command[1]))
                "down" -> Down(parseLong(command[1]))
                "up" -> Up(parseLong(command[1]))
                else -> throw IllegalArgumentException("$command")
            }
        }.fold(Position(0L, 0L)) { position, command ->
            command.move(position)
        }
        return finalPosition.position * finalPosition.depth
    }

    fun part2(input: List<String>): Long {
        val finalPosition = input.map { line ->
            val command = line.split(" ")
            when (command.first()) {
                "forward" -> Forward(parseLong(command[1]))
                "down" -> Down(parseLong(command[1]))
                "up" -> Up(parseLong(command[1]))
                else -> throw IllegalArgumentException("$command")
            }
        }.fold(PositionWithAim(0L, 0L, 0L)) { position, command ->
            command.moveAimed(position)
        }
        return finalPosition.position * finalPosition.depth
    }

    // test if implementation meets criteria from the description, like:
    val testInput = readInput("Day02_test")
    check(part1(testInput) == 150L)
    check(part2(testInput) == 900L)

    val input = readInput("Day02")
    println(part1(input))
    println(part2(input))
}

File: rust/src/day02.rs
struct Position {
    position: u32,
    depth: u32,
}

impl Position {
    fn new(position: u32, depth: u32) -> Self {
        Position {
            position: position,
            depth: depth,
        }
    }
}

struct PositionWithAim {
    position: u32,
    depth: u32,
    aim: u32,
}

impl PositionWithAim {
    fn new(position: u32, depth: u32, aim: u32) -> Self {
        PositionWithAim {
            position: position,
            depth: depth,
            aim: aim,
        }
    }
}

enum Command {
    Forward(u32),
    Up(u32),
    Down(u32),
}

#[aoc_generator(day2)]
fn convert_lines(input: &str) -> Vec<Command> {
    String::from(input).split("\n")
        .map(|line| {
            if let Some((cmd_type, magnitude)) = line.split_once(" ") {
                match cmd_type {
                    "forward" => Command::Forward(magnitude.parse().unwrap()),
                    "up" => Command::Up(magnitude.parse().unwrap()),
                    "down" => Command::Down(magnitude.parse().unwrap()),
                    _ => panic!(),
                }
            } else {
                panic!()
            }
        })
        .collect()
}

#[aoc(day2, part1)]
fn solution1(data: &[Command]) -> u32 {
    let final_position = data
        .clone()
        .iter()
        .fold(Position::new(0, 0), |position, command| match command {
            Command::Forward(magnitude) => {
                Position::new(position.position + magnitude, position.depth)
            }
            Command::Up(magnitude) => Position::new(position.position, position.depth - magnitude),
            Command::Down(magnitude) => {
                Position::new(position.position, position.depth + magnitude)
            }
        });
    final_position.position * final_position.depth
}

#[aoc(day2, part2)]
fn solution2(data: &[Command]) -> u32 {
    let final_position = data.clone().iter().fold(
        PositionWithAim::new(0, 0, 0),
        |position, command| match command {
            Command::Forward(magnitude) => PositionWithAim::new(
                position.position + magnitude,
                position.depth + (position.aim * magnitude),
                position.aim,
            ),
            Command::Up(magnitude) => {
                PositionWithAim::new(position.position, position.depth, position.aim - magnitude)
            }
            Command::Down(magnitude) => {
                PositionWithAim::new(position.position, position.depth, position.aim + magnitude)
            }
        },
    );
    final_position.position * final_position.depth
}

